### PR TITLE
[compiler] Add SJavaString / SJavaBytes / SJavaArrayString

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala [hunk 1]
```diff
@@ -266,14 +257,16 @@ object StringFunctions extends RegistryFunctions {
...
-            val m = v1.loadLength().cne(v2.loadLength())
+            val l1 = cb.newLocal[Int]("hamming_len_1", v1.invoke[Int]("length"))
+            val l2 = cb.newLocal[Int]("hamming_len_2", v2.invoke[Int]("length"))
+            val m = l1.cne(l2)
 
             IEmitCode(cb, m, {
-              cb.whileLoop(i < v1.loadLength(), {
-                cb.ifx(v1.loadByte(i).cne(v2.loadByte(i)),
+              cb.whileLoop(i < l1, {
+                cb.ifx(v1.invoke[Int, Char]("charAt", i).toI.cne(v2.invoke[Int, Char]("charAt", i).toI),
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala:269]: This is a behavioral change. The old version calculated the hamming distance on bytes this one does so on character sequences.

File: hail/src/main/scala/is/hail/expr/ir/orderings/ShuffleOrdering.scala [hunk 2]
```diff
@@ -19,7 +19,7 @@ object ShuffleOrdering {
             val bcode1 = x.asInstanceOf[SCanonicalShufflePointerCode].binaryRepr
             val bcode2 = y.asInstanceOf[SCanonicalShufflePointerCode].binaryRepr
             val ord = BinaryOrdering.make(bcode1.st, bcode2.st, ecb)
-            ord.compareNonnull(cb, x.asString.asBytes(), y.asString.asBytes())
+            ord.compareNonnull(cb, x.asInstanceOf[SCanonicalShufflePointerCode].binaryRepr, y.asInstanceOf[SCanonicalShufflePointerCode].binaryRepr)
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/orderings/ShuffleOrdering.scala:22]: can't this just use `bcode1` and `bcode2`?

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SJavaBytes.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+class SJavaBytesCode(val bytes: Code[Array[Byte]]) extends SBinaryCode {
+  def st: SBinary = SJavaBytes
+
+  def makeCodeTuple(cb: EmitCodeBuilder): IndexedSeq[Code[_]] = FastIndexedSeq(bytes)
+
+  def loadLength(): Code[Int] = bytes.invoke[Int]("length")
+
+  private[this] def memoizeWithBuilder(cb: EmitCodeBuilder, name: String, sb: SettableBuilder): SBinaryValue = {
+    val s = new SJavaBytesSettable(sb.newSettable[Array[Byte]]("sjavabytes_memoize"))
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SJavaBytes.scala:48]: `name` is unused, maybe need to use `SJavaBytesSettable.apply`?

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SJavaString.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  def loadLength(): Code[Int] = s.invoke[Int]("length")
+
+  def loadString(): Code[String] = s
+
+  def toBytes(): SBinaryCode = {
+    new SJavaBytesCode(s.invoke[Array[Byte]]("getBytes"))
+  }
+
+  private[this] def memoizeWithBuilder(cb: EmitCodeBuilder, name: String, sb: SettableBuilder): SValue = {
+    val s = new SJavaStringSettable(sb.newSettable[String]("sjavastring_memoize"))
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SJavaString.scala:60]: `name` unused, same as `SJavaBytes` comment

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 5]
```diff
@@ -444,23 +441,41 @@ abstract class RegistryFunctions {
...
+    def wrap(cb: EmitCodeBuilder, r: Value[Region], code: SCode): Code[_] = code.st.virtualType match {
       case t if t.isPrimitive => SType.extractPrimCode(cb, code)
-      case call: SCall => code.asCall.loadCanonicalRepresentation(cb)
+      case TCall => code.asCall.loadCanonicalRepresentation(cb)
+      case TArray(TString) => code.st match {
+        case _: SJavaArrayString => code.asInstanceOf[SJavaArrayStringCode].array
+        case _ =>
+          val sv = code.asIndexable.memoize(cb, "scode_array_string")
+          val arr = cb.newLocal[Array[String]]("scode_array_string", Code.newArray[String](sv.loadLength()))
+          sv.foreach(cb) { case (cb, idx, elt) =>
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:459]: why do you need this new `foreach` method rather than `foreachDefined` here?

**Review by chrisvittal — CHANGES_REQUESTED** [review]
Good work. Found a few minor issues.

File: hail/src/main/scala/is/hail/expr/ir/orderings/ShuffleOrdering.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/orderings/ShuffleOrdering.scala:22]: oh, yeah.

File: hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala:269]: I agree, the old implementation was wrong.

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SJavaBytes.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SJavaBytes.scala:48]: just need to use `name` in the string passed here.

File: hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala (hunk 5 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala:459]: ah, that will work.

**Review by chrisvittal — APPROVED** [review]
🚀 

now test batch needs to pass

**Pull request merged**